apps/hannk/interpreter/lower.cpp

```diff
@@ -74,4 +74,112 @@ OpPtr lower_tflite_lstm(TensorPtr data_input, TensorPtr prev_activ_input, Tensor
     return make_op<OpGroup>(std::move(inputs), std::move(outputs), std::move(ops));
 }
 
+namespace {
+
+TensorPtr make_shape_tensor(const TensorPtr &t) {
+    const auto &b = t->buffer();
+    const int dims = b.dimensions();
+    HalideBuffer<int32_t> data(dims);
+    for (int d = 0; d < dims; d++) {
+        data.data()[d] = b.extent(dims - d - 1);
+    }
+    TensorPtr shape_tensor = std::make_shared<Tensor>(t->name() + ".shape_tensor", std::move(data));
+    shape_tensor->set_constant();
+    return shape_tensor;
+}
+
+struct TensorAndOp {
+    TensorPtr tensor;
+    OpPtr op;
+};
+
+TensorAndOp reshape_tensor(TensorPtr t, bool is_output = false) {
+    assert(t->is_dense());
+    assert(t->rank() == 2 || t->rank() == 4);
+
+    TensorPtr t_reshaped;
+    std::string name = t->name() + ".reshaped";
+    if (t->is_constant() && !t->is_external() && !t->is_dynamic() && !t->is_alias()) {
+        // We don't need to use the normal aliasing approach here: the source
+        // is a constant-external Tensor, so we can just make a new Tensor that uses the same buffer,
+        // and thus the same underlying data, but with extra dimensions inserted.
+        HalideBuffer<void> nb = t->buffer();
+        if (t->rank() == 2) {
+            nb.add_dimension();
+            nb.add_dimension();
+            nb.transpose(1, 3);
+        }
+
+        // Reshape (c, x, y, b) to (cxy, 1, 1, b)
+        auto *dim = nb.raw_buffer()->dim;
+        dim[0].extent *= dim[1].extent * dim[2].extent;
+        dim[1].extent = 1;
+        dim[2].extent = 1;
+
+        // Canonicalize the strides as dense.
+        int stride = 1;
+        for (int i = 0; i < 4; i++) {
+            dim[i].stride = stride;
+            stride *= dim[i].extent;
+        }
+
+        t_reshaped = std::make_shared<Tensor>(name, std::move(nb), t->quantization());
+    } else {
+        Box bounds = t->bounds();
+#ifndef NDEBUG
+        for (const auto &i : bounds) {
+            assert(i.min == 0);
+        }
+#endif
+        int c_extent = bounds.front().extent();
+        int b_extent = bounds.back().extent();
+        if (bounds.size() == 4) {
+            // Reshape (c, x, y, b) to (cxy, 1, 1, b)
+            c_extent = bounds[0].extent() * bounds[1].extent() * bounds[2].extent();
+        }
+        Box reshaped_bounds = {{0, c_extent - 1}, {0, 0}, {0, 0}, {0, b_extent - 1}};
+        t_reshaped = std::make_shared<Tensor>(name, t->type(), std::move(reshaped_bounds), t->quantization());
+    }
+    t_reshaped->set_constant(t->is_constant());
```

Comment:
I don't think this should be set. Constant folding should do it if possible. I think we don't want to mark something constant before it's actually computed like this.

Reply:
Done.